src/main/java/org/reactome/release/common/dataretrieval/FileRetriever.java

```diff
@@ -0,0 +1,364 @@
+		{
+			String errorString = "5xx reply code detected (" + client.getReplyCode() + "), reply string is: "+client.getReplyString();
+			logger.error(errorString);
+			throw new Exception(errorString);
+		}
+		client.logout();
+		client.disconnect();
+	}
+
+	protected void writeInputStreamToFile(InputStream inStream) throws IOException, FileNotFoundException
+	{
+		try(ByteArrayOutputStream baos = new ByteArrayOutputStream())
+		{
+			int b = inStream.read();
+			while (b!=-1)
+			{
+				baos.write(b);
+				b = inStream.read();
+			}
+	
+			try (FileOutputStream file = new FileOutputStream(this.destination))
+			{
+				baos.writeTo(file);
+				file.flush();
+			}
+		}
+	}
+
+	
+	protected void doHttpDownload(Path path) throws HttpHostConnectException, IOException, Exception
+	{
+		this.doHttpDownload(path, HttpClientContext.create());
+	}
+	
+	protected void doHttpDownload(Path path, HttpClientContext context) throws Exception, HttpHostConnectException, IOException
+	{
+		HttpGet get = new HttpGet(this.uri);
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		RequestConfig config = RequestConfig.copy(RequestConfig.DEFAULT)
+											.setConnectTimeout(1000 * (int)this.timeout.getSeconds())
+											.setSocketTimeout(1000 * (int)this.timeout.getSeconds())
+											.setConnectionRequestTimeout(1000 * (int)this.timeout.getSeconds()).build();
+		
+		get.setConfig(config);
+		
+		int retries = this.numRetries;
+		boolean done = retries + 1 <= 0;
+		while(!done)
+		{
+			try( CloseableHttpClient client = HttpClients.createDefault();
+				CloseableHttpResponse response = client.execute(get, context);
+				OutputStream outputFile = new FileOutputStream(path.toFile()))
+			{
+				int statusCode = response.getStatusLine().getStatusCode();
+				// If status code was not 200, we should print something so that the users know that an unexpected response was received.
+				if (statusCode != HttpStatus.SC_OK)
+				{
+					if (String.valueOf(statusCode).startsWith("4") || String.valueOf(statusCode).startsWith("5"))
+					{
+						logger.error("Response code was 4xx/5xx: {}, Status line is: {}", statusCode, response.getStatusLine());
+					}
+					else
+					{
+						logger.warn("Response was not \"200\". It was: {}", response.getStatusLine());
+					}
+				}
+				response.getEntity().writeTo(outputFile);
+				done = true;
+			}
+			catch (ConnectTimeoutException e)
+			{
+				// we will only be retrying the connection timeouts, defined as the time required to establish a connection.
+				// we will not handle socket timeouts (inactivity that occurs after the connection has been established).
+				// we will not handle connection manager timeouts (time waiting for connection manager or connection pool).
+				e.printStackTrace();
+				logger.info("Failed due to ConnectTimeout, but will retry {} more time(s).", retries);
+				retries--;
+				done = retries + 1 <= 0;
```

Comment:
Any reason why this is evaluated again, and within the catch block?

Reply:
I'm not sure I understand the issue here.

Reply:
Ah nevermind, I think I was confused about the boolean being updated/checked within the catch block. Looking over it again this morning, I get it